File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int cnt = 0;
    void dfs(TreeNode* node, int maxVal) {
        if (node == nullptr) {
            return;
        }
        if (node->val >= maxVal) {
            cnt++;
            maxVal = node->val;
        }
        dfs(node->left, maxVal);
        dfs(node->right, maxVal);
    }
public:
    int goodNodes(TreeNode* root) {
        dfs(root, INT_MIN);
        return cnt;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int maxDiameter = 0;
    private int maxDepth(TreeNode node) {
        if (node == null) {
            return 0;
        }
        maxDiameter = Math.max(maxDepth(node.left) + maxDepth(node.right), maxDiameter);
        return Math.max(maxDepth(node.left) + 1, maxDepth(node.right) + 1);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        maxDiameter = this.maxDiameter;
        maxDepth(root);
        return maxDiameter;
    }
}
